full_bubble_model.py

```diff
@@ -361,91 +565,166 @@ def get_data(cpar, num_sol, error_code, elapsed_time):
+    data[f'n_{target_specie}'] = 0.0
+    data.m_target = 0.0
+    data.expansion_work = 0.0
+    data.dissipated_acoustic_energy = 0.0
+    data.energy_efficiency = 1.0e30
+    data.enable_heat_transfer = enable_heat_transfer
+    data.enable_evaporation = enable_evaporation
+    data.enable_reactions = enable_reactions
+    data.enable_dissipated_energy = enable_dissipated_energy
+    data.excitation_type = excitation_type
+    data.target_specie = target_specie
+    errors, success = get_errors(error_code)
+    data.success = success
+    if not success:
+        return data
     
     # normal functioning
-    else:
-        data.steps = len(num_sol.t)
+    data.steps = len(num_sol.t)
+    data.x_initial = num_sol.y[:, 0] # initial values of [R, R_dot, T, c_1, ... c_K]
         
-        # collapse time (first loc min of R)
-        loc_min = argrelmin(num_sol.y[:][0])
-        data.collapse_time = 0.0
-        if not len(loc_min[0]) == 0:
-            data.collapse_time = num_sol.t[loc_min[0][0]]
+    # collapse time (first loc min of R)    TODO fix
+    loc_min = argrelmin(num_sol.y[:][0])
+    data.collapse_time = 0.0
+    if not len(loc_min[0]) == 0:
+        data.collapse_time = num_sol.t[loc_min[0][0]]
         
-        # Energy calculations
-        data.T_max = np.max(num_sol.y[:][2]) # maximum of temperature peaks [K]
-        data.x_final = num_sol.y[:, -1] # final values of [R, R_dot, T, c_1, ... c_K]
-        last_V = 4.0 / 3.0 * (100.0 * data.x_final[0]) ** 3 * np.pi # [cm^3]
-        data.n_H2 = data.x_final[3+par.index['H2']] * last_V # [mol]
-        data.n_O2 = data.x_final[3+par.index['O2']] * last_V # [mol]
-        m_H2 = 1e-3 * data.n_H2 * par.W[par.index['H2']] # [kg]
-        data.work = Work(cpar) # [J]
-        data.energy = 1e-6 * data.work / m_H2 # [MJ/kg]
-    
+    # Energy calculations
+    data.T_max = np.max(num_sol.y[:][2]) # maximum of temperature peaks [K]
+    data.x_final = num_sol.y[:, -1] # final values of [R, R_dot, T, c_1, ... c_K]
+    last_V = 4.0 / 3.0 * (100.0 * data.x_final[0]) ** 3 * np.pi # [cm^3]
+    data[f'n_{target_specie}'] = data.x_final[3+par.index[target_specie]] * last_V # [mol]
+    m_target = 1.0e-3 * data[f'n_{target_specie}'] * par.W[par.index[target_specie]] # [kg]
+    data.expansion_work = Work(cpar, enable_evaporation) # [J]
+    data.dissipated_acoustic_energy = data.x_final[-1]  # [J]
+    all_work = data.expansion_work + data.dissipated_acoustic_energy
+    data.energy_efficiency = 1.0e-6 * all_work / m_target if m_target > 0.0 else 1.0e30 # [MJ/kg]
+    data.target_specie = target_specie
     return data
 
-
 # keys of data: (except x_final)
-keys = ['ID', 'R_E', 'ratio', 'P_inf', 'alfa_M', 'T_inf', 'P_v', 'mu_L', 'gases', 'fractions', 'surfactant', 'error_code', 'elapsed_time', 'steps', 'collapse_time', 'T_max', 'n_H2', 'n_O2', 'work', 'energy']
+keys = ['ID', 'R_E', 'ratio', 'P_amb', 'alfa_M', 'T_inf', 'P_v', 'mu_L', 'gases', 'fractions', 'surfactant', 'c_L',
```

Comment:
more info is saved, like settings, excitation type and args, success, dissipated_energy. Target specie (H2 or NH3) have to be specified in settings